hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpContractStoresSubcommand.java

```diff
@@ -167,28 +196,125 @@ public static ContractKeyLocal from(ContractKey ckey) {
+
+        if (contractStorageStore == null) return false;
+        contractStorageStore.keys().forEachRemaining(key -> {
+            // (Not sure how many temporary _copies_ of a byte arrays are made here ... best not to ask ...)
+            final var contractKeyLocal = ContractKeyLocal.from(
+                    new ContractKey(key.contractNumber(), key.key().toByteArray()));
+            final var slotValue = contractStorageStore.get(key);
+            assert (slotValue != null);
+            final var value = uint256FromByteArray(slotValue.value().toByteArray());
+            nSlotsSeen.incrementAndGet();
+            visitor.accept(contractKeyLocal, value);
+        });
+
+        System.err.printf("iterateThroughContractStorage: end - %d slots seen, completed? true%n", nSlotsSeen.get());
+
+        return true;
+    }
+
+    /** First migrates the contract store from the mono-service representation to the modular-service representations,
+     *  and then returns all contracts with bytecodes from the migrated contract store, plus the ids of contracts with
+     *  0-length bytecodes.
+     */
+    @SuppressWarnings("unchecked")
+    @Nullable
+    WritableKVState<SlotKey, SlotValue> getMigratedContractStore() {
+
+        final var fromStore = state.getRawContractStorage();
+        final var expectedNumberOfSlots = Math.toIntExact(fromStore.size());
+
+        // Start the migration with a clean, writable KV store.  Using the in-memory store here.
+
+        final var contractSchema = new ContractSchema();
+        final var contractSchemas = contractSchema.statesToCreate();
+        final StateDefinition<SlotKey, SlotValue> contractStoreStateDefinition = contractSchemas.stream()
+                .filter(sd -> sd.stateKey().equals(ContractSchema.STORAGE_KEY))
+                .findFirst()
+                .orElseThrow();
+        final var contractStoreSchemaMetadata =
+                new StateMetadata<>(ContractService.NAME, contractSchema, contractStoreStateDefinition);
+        final var contractMerkleMap =
+                new NonAtomicReference<MerkleMap<InMemoryKey<SlotKey>, InMemoryValue<SlotKey, SlotValue>>>(
+                        new MerkleMap<>(expectedNumberOfSlots));
+        final var toStore = new NonAtomicReference<WritableKVState<SlotKey, SlotValue>>(
+                new InMemoryWritableKVState<>(contractStoreSchemaMetadata, contractMerkleMap.get()));
+
+        final var flushCounter = new AtomicInteger();
+
+        final ContractStateMigrator.StateFlusher stateFlusher = ignored -> {
+            // Commit all the new leafs to the underlying map
+            ((WritableKVStateBase<SlotKey, SlotValue>) (toStore.get())).commit();
+            // Copy the underlying map, which does the flush
+            contractMerkleMap.set(contractMerkleMap.get().copy());
+            // Create a new store to go on with
+            toStore.set(new InMemoryWritableKVState<>(contractStoreSchemaMetadata, contractMerkleMap.get()));
```

Comment:
What advantage does using a new object give us?

Reply:
(BTW the use of the `NonAtomicReferences` is just to ~fake out~ take proper advantage of Java's idea of needing "final" (or "effectively final") things for lambda captures.)